magic_items/services.py

```diff
@@ -0,0 +1,358 @@
+
+        item_model = MagicItemModel.query.get_or_404(magic_item_id)
+
+        for k, v in form.data.items():
+            if k != CSRF_TOKEN and k != DESCRIPTION:
+                setattr(item_model, k, v)
+
+        item_model.description = (
+            f"{form.rarity.data}, {form.item_type.data}",
+            form.description.data,
+        )
+        item_model.last_updated = datetime.utcnow()
+
+        db.session.commit()
+
+        return MagicItemDomain.from_model(item_model)
+
+    @classmethod
+    def delete(cls, user_id: int, magic_item_id: int) -> bool:
+
+        item_model = MagicItemModel.query.get_or_404(magic_item_id)
+
+        if item_model.created_by is not user_id:
+            abort(403)
+
+        db.session.delete(item_model)
+        db.session.commit()
+        return True
+
+    @classmethod
+    def random(
+        cls,
+    ) -> MagicItemDomain:
+
+        item_model = (
+            MagicItemModel.query.options(load_only("id"))
+            .offset(
+                func.floor(
+                    func.random() * db.session.query(func.count(MagicItemModel.id))
+                )
+            )
+            .limit(1)
+            .all()[0]
+        )
+
+        return MagicItemDomain.from_model(item_model)
+
+
+class CollectionService:
+    @classmethod
+    def create(
+        cls,
+        user_id: int,
+        form: CollectionAddForm,
+    ) -> CollectionDomain:
+
+        data = {k: v for k, v in form.data.items() if k != CSRF_TOKEN}
+
+        collection_model = CollectionModel(
+            **data,
+            user_id=user_id,
+        )
+
+        db.session.add(collection_model)
+        db.session.commit()
+
+        return CollectionDomain.from_model(collection_model)
+
+    @classmethod
+    def update(
+        cls,
+        user_id: int,
+        collection_id: int,
+        form: CollectionAddForm,
+    ) -> CollectionDomain:
+
+        collection_model = CollectionModel.query.get_or_404(collection_id)
+
+        if collection_model.user_id is not user_id:
+            abort(403)
+
+        for k, v in form.data.items():
+            if k != CSRF_TOKEN:
+                setattr(collection_model, k, v)
+
+        collection_model.last_updated = datetime.utcnow()
+
+        db.session.commit()
+
+        return CollectionDomain.from_model(collection_model)
+
+    @classmethod
+    def delete(cls, collection_id: int, user_id: int) -> bool:
+
+        collection_model = CollectionModel.query.get(collection_id)
+
+        if collection_model.user_id is not user_id:
+            abort(403)
+
+        db.session.delete(collection_model)
+        db.session.commit()
+        return True
+
+    @classmethod
+    def get(
+        cls,
+        collection_id: int,
+    ) -> CollectionDomain:
+        collection_model = CollectionModel.query.get(collection_id)
+        return CollectionDomain.from_model(collection_model)
+
+    @classmethod
+    def get_filtered(
+        cls,
+        **filters: dict,
```

Comment:
One catch with using the ** here like this is you can't control what kind of filters are being accepted. A pattern that can be used is to have a domain class like StandardCollectionFilters where you specify acceptable filter fields/values.